contoso_manufacturing/developer/webapp-decode/yolov8.py

```diff
@@ -156,22 +185,41 @@ def draw_detections(self, img, box, score, class_id):
         cv2.putText(img, label, (label_x, label_y), cv2.FONT_HERSHEY_SIMPLEX, 1.2, (0, 0, 0), 1, cv2.LINE_AA)
 
     def run(self):
-        if(self.verbose):
-            print("Running detection...")
+        while not self.stopped:
+            if(self.verbose):
+                print("Running detection...")
+
+            self.frame_number += 1
+            # If mod = 0, i will get the frame and skip it
+            if self.frame_number % self.skip_rate == 0:
+                if not self.frame_queue.empty():
+                    self.cap = self.frame_queue.get()
+                continue
+            
+            if not self.frame_queue.empty():
+                frame = self.frame_queue.get()
+                # Preprocess the current frame
+                image_data = self.preprocess(frame)
 
-        self.frame_number += 1
-        # If mod = 0, i will get the frame and skip it
-        if self.frame_number % self.skip_rate == 0:
-            self.cap.read()
-            return None
-        
-        image_data = self.preprocess()
+                # Send the preprocessed frame to the gRPC client for prediction
+                outputs = self.grpc_client.predict({"images": image_data}, self.model_name)
+
+                # Postprocess the prediction results and get the final frame
+                processed_frame = self.postprocess(frame, outputs)
+
+                return processed_frame
+        # image_data = self.preprocess()
 
-        outputs = self.grpc_client.predict({"images": image_data}, self.model_name)
-        frame = self.postprocess(self.cap.read()[1], outputs)
+        # outputs = self.grpc_client.predict({"images": image_data}, self.model_name)
+        # frame = self.postprocess(self.cap.read()[1], outputs)
 
-        return frame
+        # return frame
 
+    def stop(self):
+        with self.lock:
```

Comment:
Can you include a code comment to explain what this code is doing (why the lock is needed)?